File: components/maps/src/main/java/org/storm/maps/layer/BackgroundLayer.java
package org.storm.maps.layer;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.image.Image;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.storm.core.ui.Resolution;
import org.storm.core.utils.ImageUtils;
import org.storm.maps.exception.MapLayerException;
import org.storm.physics.entity.Entity;

import java.io.FileInputStream;
import java.io.FileNotFoundException;

@Setter
public class BackgroundLayer extends Layer {

  private final Image background;

  public BackgroundLayer(Image background, Resolution resolution) {
    super(false, resolution);
    this.background = background;
  }

  public BackgroundLayer(String imagePath, Resolution resolution) throws FileNotFoundException {
    this(new Image(new FileInputStream(imagePath)), resolution);
  }

  @Override
  public void addEntity(Entity entity) {
    throw new MapLayerException("cannot add entities to the background layer");
  }

  @Override
  public void removeEntity(Entity entity) {
    throw new MapLayerException("cannot remove entities to the background layer");
  }

  @Override
  public void render(GraphicsContext graphicsContext, double x, double y) {
    Image croppedBackground = ImageUtils.crop(this.background, (int)x, (int)y,
      (int)(x + this.resolution.getWidth()), (int)(y + this.resolution.getHeight()));
    graphicsContext.drawImage(croppedBackground, 0, 0);
  }

  @Override
  public void update(double time, double elapsedTime) {
    // Default behaviour involves no updating but something like Parallax might involve the need to update
  }
}

File: components/maps/src/main/kotlin/org/storm/maps/layer/BackgroundLayer.kt
package org.storm.maps.layer

import org.storm.core.context.Context
import org.storm.core.context.RESOLUTION
import org.storm.core.render.canvas.Canvas
import org.storm.core.render.Image
import org.storm.core.render.impl.JfxImage
import org.storm.maps.exception.MapLayerException
import org.storm.physics.entity.Entity

class BackgroundLayer(
    private val background: Image
) : Layer(false) {

    constructor(imagePath: String) : this(JfxImage(imagePath))

    override fun addEntity(entity: Entity) {
        throw MapLayerException("cannot add entities to the background layer")
    }

    override fun removeEntity(entity: Entity) {
        throw MapLayerException("cannot remove entities to the background layer")
    }

    override suspend fun render(canvas: Canvas, x: Double, y: Double) {
        val resolution = Context.RESOLUTION
        val croppedBackground = background.crop(x, y, x + resolution.width, y + resolution.height)
        canvas.drawImageWithPixels(croppedBackground, 0.0, 0.0)
    }

    override suspend fun update(time: Double, elapsedTime: Double) {
        // Default behaviour involves no updating but something like Parallax might involve the need to update
    }
}
